### PR TITLE
feat(pinned items): Issue #6229 - Adds pinned functionality to List Component

### PR DESCRIPTION
**Pull request by jeffcap1**
#### What does this PR do?

Adds functionality to provide a list of ItemKeys to be pinned in place within the list and will not change when the list is reordered. 

#### Where should the reviewer start?

Please review the context provided in issue 6229 explaining why this feature was requested and the proposed change that was implemented. Then review Storybook and Jest tests for further context regarding functionality and implementation. 
Finally a useMemo hook is used to derive the orderableItems from the pinnedItems which is used to replace the data object and the orderableIndex is found for each item. These replace the previously used data and index to find and set active items and updates from within the reorder function.

#### What testing has been done on this PR?

Tests have been written in Jest as well as a Storybook story. Tests verify that the pinned items are styled appropriately. Also, after a re-orderable item is moved that the pinned item does not change its position.

#### Do Jest tests follow these best practices?

- [x] `screen` is used for querying.
- [x] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [x] `userEvent` is used in place of `fireEvent`.
- [x] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?

Pleas see the context provided in Issue 6229.

#### What are the relevant issues?

[Issue #6229](https://github.com/grommet/grommet/issues/6229)

#### Screenshots (if appropriate)

image

#### Do the grommet docs need to be updated?

Yes

#### Should this PR be mentioned in the release notes?

Yes

#### Is this change backwards compatible or is it a breaking change?

The only minor change that might be breaking is regarding the ItemKey. If the PrimaryKey is provided and not the ItemKey for an object, the PrimaryKey will be set as the ItemKey. Line number is 175.
```javascript
// fixes issue where itemKey is undefined when only primaryKey is provided
const itemKey = defaultItemKey || primaryKey || null;
```

### CODE REVIEW

**Comment by jeffcap1** [comment]
It looks like importing and using the <Pin /> icon from grommet-icons has pushed the bundle size over the limit of 132kb gzipped that it's being checked by CircleCI. What is the recommended approach for using an icon? Or is there a better way to accomplish this type of styling.

**Comment by ericsoderberghp** [comment]
> It looks like importing and using the icon from grommet-icons has pushed the bundle size over the limit of 132kb gzipped that it's being checked by CircleCI. What is the recommended approach for using an icon? Or is there a better way to accomplish this type of styling.

The icon needs to be imported from the particular folder within the grommet-icons package, to avoid pulling in all icons just to dereference one of them. For example:
`import { CircleAlert } from 'grommet-icons/icons/CircleAlert';`

File: src/js/components/List/index.d.ts
Comment (ericsoderberghp) [inline src/js/components/List/index.d.ts:64]: I think this would need to be `pinned?: (string | number)[];` to align with the return value of `itemKey()`.

**Review by ericsoderberghp — COMMENTED** [review]
Just getting started reviewing

File: src/js/components/List/index.d.ts
Comment (jeffcap1) [reply src/js/components/List/index.d.ts]: Good point, I'll make that update. I originally copied the disabled type definition.

File: src/js/components/List/List.js
Comment (MikeKingdom) [inline src/js/components/List/List.js:474]: Why not allow clicking on a pinned item? That seems like the domain of disabled.

File: src/js/components/List/List.js
Comment (MikeKingdom) [inline src/js/components/List/List.js:464]: this really needs to be from the theme

File: src/js/components/List/List.js
Comment (MikeKingdom) [inline src/js/components/List/List.js:658]: Is this trying to size the icon and then add pad? Seems like you'd judy want some pad on box at 655 and a `size` on the icon.

File: src/js/components/List/List.js
Comment (MikeKingdom) [inline src/js/components/List/List.js:675]: I don't think this is doing anything.

File: .nvmrc
Comment (jeffcap1) [reply .nvmrc:1]: Okay no problem. The only reason I had added this was because I had issues running the project when I started because I was using Node 17.4. I wasn't sure if there was any benefit to standardizing on a version.

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:658]: 657 is reducing the size of the pin icon since it was too large compared to the arrow icons.

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:474]: Great point, I'll make that change.

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:675]: You're correct, I'll make sure to remove that.

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:464]: That makes total sense. I was having trouble understanding where the theme information was being set since it's coming from Styled Components. Do you know the file I should add this to or a reference page I can use as a resource?

File: src/js/components/List/List.js
Comment (MikeKingdom) [reply src/js/components/List/List.js:658]: You can use pixel sizes for the icon size property

File: src/js/components/List/List.js
Comment (MikeKingdom) [reply src/js/components/List/List.js:464]: in `src/js/themes/base.js` are the base theme definitions for grommet. Other themes like the grommet-theme-hpe theme can have a sparse object of overrides as needed. So for this, if you look at about line 1023 there is a section for `list` which is where the above code gets a couple things related to disabled list items. You could have a `pinned` part below the section for disabled that has this so something like:
```jsx
pinned: {
  color: 'background-contrast',
},
```
I didn't put `light-2' there because I don't think it's very light mode versus dark mode friendly. I think in this case `background-contrast` is more appropriate (it's mainly just an opacity difference over the existing background so it makes it darker in the light theme and lighter in the dark theme. Others may weigh in on this idea.

Comment (jeffcap1) [reply src/js/components/List/List.js:464]: Awesome thank you for the detailed information. I'll review this tonight and make the appropriate changes.

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:658]: Okay I'll give that a shot tonight. I can't remember, but I feel like I tried putting in a pixel size and it wasn't working, but I'll double check tonight. I think I just tried `size="18px"` for example, but is it supposed to an object?

Comment (MikeKingdom) [reply src/js/components/List/List.js:658]: Just a string. Here's an example:
image

Comment (MikeKingdom) [reply src/js/components/List/List.js:658]: I think you should also add the icon and the size of the icon to the theme. You can see an example in base.js for list for the up down icons on line 1030

Comment (jeffcap1) [reply src/js/components/List/List.js]: Thank you so much for all of your help. I updated the icon based on your feedback and used the base theme for that.

File: src/js/components/List/List.js
Comment (MikeKingdom) [inline src/js/components/List/List.js:660]: Do you still need two boxes here? I would think these could be combined.

**Review by MikeKingdom — COMMENTED** [review]
Looking pretty good. One small comment

File: src/js/components/List/List.js
Comment (jeffcap1) [reply src/js/components/List/List.js:660]: Ah sorry forgot to double check that. You're correct that Box is unnecessary.

**Review by MikeKingdom — APPROVED** [review]

**Comment by jeffcap1** [comment]
@ericsoderberghp Just checking if there is anything else I needed on my end for this PR to be merged.

File: src/js/components/List/List.js
Comment (ericsoderberghp) [inline src/js/components/List/List.js:463]: Since we're using this theme value for the background of the item, I think the theme path should be `theme.list.item.pinned.background`

File: src/js/components/List/List.js
Comment (ericsoderberghp) [inline src/js/components/List/List.js:646]: I think these three should be within the `if (isPinned)` below.

File: src/js/themes/base.js
Comment (ericsoderberghp) [inline src/js/themes/base.js:1035]: Can we not rely on our t-shirt sizing for these?

Comment (jeffcap1) [reply src/js/themes/base.js:1035]: @ericsoderberghp, I tried using the t-shirt sizing, but the medium is too big and small is too small in my opinion. I found that 18px was around the same size as the arrows for the non-pinned items and the 15px padding made it the same size (44px) box to keep everything consistent at all screen sizes. Below are some screenshots for reference to show the differences. Let me what you think and how you'd like me to proceed.

---

#### size: 18px
image

---

#### size: 'medium'
image

---

#### size: 'small'
image

Comment (ericsoderberghp) [reply src/js/themes/base.js:1035]: To align the icon sizes better, I think we should add a "FormPin" icon, to go with "FormUp" and "FormDown" being used. This would involve creating a new svg and adding it to `grommet-icons`.

Comment (jeffcap1) [reply src/js/themes/base.js:1035]: @ericsoderberghp, okay no problem. I've opened a [PR](https://github.com/grommet/grommet-icons/pull/283) over on the grommet-icons repo to add in this icon. Once that's merged I'll reference that in here.
Would I update the grommet-icons version in the package.json or is there a separate process to follow for that?

Comment (jcfilben) [reply src/js/themes/base.js:1035]: Hey @jeffcap1 thanks for creating the grommet-icons PR. We need to release a new version of grommet-icons so that we are able to use the new icon. I created a ticket for the team to release a new version of grommet icons that can be tracked here: https://github.com/grommet/hpe-design-system/issues/2842

Comment (jeffcap1) [reply src/js/themes/base.js:1046]: @ericsoderberghp, this has now been updated to use the new FormPin icon from the latest grommet-icons release. Please let me know if you see anything else that needs to be updated.

**Comment by jeffcap1** [comment]
@ericsoderberghp, just following up on this PR to see if there are any additional changes that need to be made before it can be merged in. Thanks!

**Comment by ericsoderberghp** [comment]
> @ericsoderberghp, just following up on this PR to see if there are any additional changes that need to be made before it can be merged in. Thanks!

Sorry for the delay on this. The code is looking good. Can you resolve the conflicts with package.json and yarn.lock? I think we're good to merge after that.

**Comment by ericsoderberghp** [comment]
It still looks like there are some conflicts with the master branch for package.json and yarn.lock

**Comment by jeffcap1** [comment]
Hey Eric. Yea I was just trying to resolve that now. I'm not getting this conflict locally so I'm going to use the Github's Resolve Conflicts screen unless you know of a better way to resolve them.

**Comment by ericsoderberghp** [comment]
> Hey Eric. Yea I was just trying to resolve that now. I'm not getting this conflict locally so I'm going to use the Github's Resolve Conflicts screen unless you know of a better way to resolve them.

I usually re-fetch/pull the master branch and merge via the command line.

**Comment by jeffcap1** [comment]
Yea that's what I tried. On my local master branch I ran `git fetch upstream` and it pull down the latest code, but to my surprise there weren't any merge conflict to resolve.

**Comment by jeffcap1** [comment]
Ah pulling upstream looks like I was able to get the merge conflict to happen. I'll have this fixed shortly.

**Comment by jeffcap1** [comment]
@ericsoderberghp, okay I think that should be fixed now

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**
